kernel/model/model.go

```diff
@@ -144,16 +149,38 @@ type Regions map[string]*Region
 
 type Region struct {
 	Scope
-	model  *Model
-	id     string
+	Model  *Model
+	Id     string
 	Region string
 	Site   string
 	Hosts  Hosts
+	Index  int
+}
+
+func (region *Region) CloneRegion(index int) *Region {
+	result := &Region{
+		Scope:  *region.CloneScope(),
+		Model:  region.Model,
+		Region: region.Region,
+		Site:   region.Site,
+		Hosts:  Hosts{},
+		Index:  index,
+	}
+	for key, host := range region.Hosts {
+		result.Hosts[key] = host.CloneHost(0)
+	}
+	return result
+}
+
+func (region *Region) Templatize(templater *Templater) {
```

Comment:
I think my biggest issue with "templatization" is the name. If all we're doing is calling `CloneXXX` methods from a `Factory`... why not just call it that? Then, doesn't the model just contain a single, or the smallest version of a footprint, and we're just cloning that to extend the model to a larger configuration?

I just don't think the model is a "template" in that example. It's a fully-fledged model.

Reply:
We can change the name, but there is more to it than just cloning. Taking the edge bit for an example, here's the client host/component definition:

```
				"client{{ .Index }}": {
					Scope: model.Scope{Tags: model.Tags{"^client", "^sdk-app"}},
					Components: model.Components{
						"client{{ .Host.Index }}": {
							BinaryName:     "ziti-fabric-test",
							PublicIdentity: "client{{ .Host.Index }}",
							ConfigSrc:      "loop/edge-perf.loop3.yml",
							ConfigName:     "edge-perf-{{ .Host.Index }}.loop3.yml",
						},
					},
				},
```
So we're not just copying it, we're using it as a base, and injecting environment into the id, public identity and config name. Even if we only had one client, it wouldn't look like the code above, it would be the version after templating has been applied. 

Does that justify the templating nomenclature, or does it still feel like it should be called something else?

Reply:
I need to spend more time thinking about this. I don't think bringing the golang template engine into the mix makes this more streamlined... I feel like it makes it more convoluted. Using the model to express things that are "templates" feels like a corruption of the model.